src/providers/wfs/qgsbackgroundcachedfeatureiterator.cpp

```diff
@@ -0,0 +1,870 @@
+        {
+          if ( !mRequest.filterFids().contains( feat.id() ) )
+          {
+            continue;
+          }
+        }
+
+        QgsGeometry constGeom = feat.geometry();
+        if ( !mFilterRect.isNull() &&
+             ( constGeom.isNull() || !constGeom.intersects( mFilterRect ) ) )
+        {
+          continue;
+        }
+
+        copyFeature( feat, f, false );
+        return true;
+      }
+
+      // When the stream is finished, cleanup
+      cleanupReaderStreamAndFile();
+
+      // And try again to check if there's a new output stream to read from
+      continue;
+    }
+
+    if ( mDownloadFinished )
+      return false;
+    if ( mInterruptionChecker && mInterruptionChecker->isCanceled() )
+      return false;
+
+    //QgsDebugMsg("fetchFeature before loop");
+    QEventLoop loop;
+    mLoop = &loop;
+    QTimer timer( this );
+    timer.start( 50 );
+    QTimer requestTimeout( this );
+    if ( mRequest.timeout() > 0 )
+    {
+      connect( &requestTimeout, &QTimer::timeout, this, &QgsBackgroundCachedFeatureIterator::timeout );
+      requestTimeout.start( mRequest.timeout() );
+    }
+    if ( mInterruptionChecker )
+      connect( &timer, &QTimer::timeout, this, &QgsBackgroundCachedFeatureIterator::checkInterruption );
+    loop.exec( QEventLoop::ExcludeUserInputEvents );
+    mLoop = nullptr;
+    //QgsDebugMsg("fetchFeature after loop");
+  }
+}
+
+void QgsBackgroundCachedFeatureIterator::cleanupReaderStreamAndFile()
+{
+  if ( mReaderStream )
+  {
+    delete mReaderStream;
+    mReaderStream = nullptr;
+    delete mReaderFile;
+    mReaderFile = nullptr;
+    mReaderByteArray.clear();
+    if ( !mReaderFilename.isEmpty() )
+    {
+      QFile::remove( mReaderFilename );
+      mReaderFilename.clear();
+      mShared->releaseCacheDirectory();
+    }
+  }
+}
+
+bool QgsBackgroundCachedFeatureIterator::rewind()
+{
+  if ( mClosed )
+    return false;
+
+  cleanupReaderStreamAndFile();
+
+  QgsFeatureRequest requestCache;
+  int genCounter = mShared->getUpdatedCounter();
+  if ( genCounter >= 0 )
+    requestCache.setFilterExpression( QString( QgsBackgroundCachedFeatureIteratorConstants::FIELD_GEN_COUNTER + " <= %1" ).arg( genCounter ) );
+  else
+    mDownloadFinished = true;
+  auto cacheDataProvider = mShared->cacheDataProvider();
+  if ( cacheDataProvider )
+    mCacheIterator = cacheDataProvider->getFeatures( requestCache );
+  return true;
+}
+
+bool QgsBackgroundCachedFeatureIterator::close()
+{
+  if ( mClosed )
+    return false;
+  QgsDebugMsgLevel( QStringLiteral( "QgsBackgroundCachedFeatureIterator::close()" ), 4 );
+
+  iteratorClosed();
+
+  mClosed = true;
+  return true;
+}
+
+
+void QgsBackgroundCachedFeatureIterator::copyFeature( const QgsFeature &srcFeature, QgsFeature &dstFeature, bool srcIsCache )
+{
+  //copy the geometry
+  QgsGeometry geometry = srcFeature.geometry();
+  if ( mShared->hasGeometry() && !geometry.isNull() )
+  {
+    QgsGeometry g;
+    g.fromWkb( geometry.asWkb() );
```

Comment:
why not just set directly? You'll get the benefits of the implicit sharing and a near-zero cost copy...

Reply:
taken into account. I was a bit worried about the thread-safety of the sharing (which could be an issue here), but looking at QgsGeometry code, it looks OK with respect to that. Taken into account